### ISSUE TITLE
(Android/FlatList) Fix onMomentumScrollEnd being called multiple times

### ISSUE THREAD

**Issue by huangkaiyuan** [facebook/react-native#32696]
### Description

![image](https://user-images.githubusercontent.com/36758804/144400195-53fcd793-be01-4466-9d54-c5a21a8f3512.png)

### Version

0.64.1

### Output of `react-native info`

![image](https://user-images.githubusercontent.com/36758804/144400195-53fcd793-be01-4466-9d54-c5a21a8f3512.png)

### Steps to reproduce

![image](https://user-images.githubusercontent.com/36758804/144400195-53fcd793-be01-4466-9d54-c5a21a8f3512.png)

**Comment by fanyangxi** [comment]
Observing the same symptom.
FYI, the `onMomentumScrollBegin` was triggered only once, but the `onMomentumScrollEnd` was trigged around 3~5 times.

![image](https://user-images.githubusercontent.com/6715050/145373045-09b27f91-3694-45b8-87e2-36230851b287.png)

**Comment by Arjan-Zuidema** [comment]
Same bug occurs on ScrollView

**Comment by meowcorp** [comment]
+1

**Comment by vamper424** [comment]
Muchas gracias, estudiaré las acciones github mi amigo.

**Comment by anton-patrushev** [comment]
Was able to reproduce it on the 0.66.4 version. Android only. Hermes is enabled (maybe it would be useful).
I have faced with that issue on 0.64.2 first time. But it's still reproducible.

**Comment by rubydeve** [comment]
+1

**Comment by rubydeve** [comment]
This is how it worked for me
```
const [canmomentum, setCanMomentum] = useState(false);
<ScrollView
      onScroll={( event ) => {
           setCanMomentum(true)
     }}
     onMomentumScrollEnd={() => {
           if (canmomentum) console.log('onMomentumScrollEnd')
           setCanMomentum(false)
     }}
>
     <Content/>
</ScrollView>
```

**Comment by ewnavilae** [comment]
> This is how it worked for me
> 
> ```
> const [canmomentum, setCanMomentum] = useState(false);
> <ScrollView
>       onScroll={( event ) => {
>            setCanMomentum(true)
>      }}
>      onMomentumScrollEnd={() => {
>            if (canmomentum) console.log('onMomentumScrollEnd')
>            setCanMomentum(false)
>      }}
> >
>      <Content/>
> </ScrollView>
> ```

Thanks for this suggestion, I did something similar but with a ref instead to avoid unnecessary state updates and renders.

**Comment by achiyazigi** [comment]
> > This is how it worked for me
> > ```
> > const [canmomentum, setCanMomentum] = useState(false);
> > <ScrollView
> >       onScroll={( event ) => {
> >            setCanMomentum(true)
> >      }}
> >      onMomentumScrollEnd={() => {
> >            if (canmomentum) console.log('onMomentumScrollEnd')
> >            setCanMomentum(false)
> >      }}
> > >
> >      <Content/>
> > </ScrollView>
> > ```
> 
> Thanks for this suggestion, I did something similar but with a ref instead to avoid unnecessary state updates and renders.

I didn't find how to reference the scrolling state from ref. would you agree to share your solution?

**Comment by asr1191** [comment]
> > This is how it worked for me
> > ```
> > const [canmomentum, setCanMomentum] = useState(false);
> > <ScrollView
> >       onScroll={( event ) => {
> >            setCanMomentum(true)
> >      }}
> >      onMomentumScrollEnd={() => {
> >            if (canmomentum) console.log('onMomentumScrollEnd')
> >            setCanMomentum(false)
> >      }}
> > >
> >      <Content/>
> > </ScrollView>
> > ```
> 
> Thanks for this suggestion, I did something similar but with a ref instead to avoid unnecessary state updates and renders.



> > > This is how it worked for me
> > > ```
> > > const [canmomentum, setCanMomentum] = useState(false);
> > > <ScrollView
> > >       onScroll={( event ) => {
> > >            setCanMomentum(true)
> > >      }}
> > >      onMomentumScrollEnd={() => {
> > >            if (canmomentum) console.log('onMomentumScrollEnd')
> > >            setCanMomentum(false)
> > >      }}
> > > >
> > >      <Content/>
> > > </ScrollView>a
> > > ```
> > 
> > 
> > Thanks for this suggestion, I did something similar but with a ref instead to avoid unnecessary state updates and renders.
> 
> I didn't find how to reference the scrolling state from ref. would you agree to share your solution?



> > > This is how it worked for me
> > > ```
> > > const [canmomentum, setCanMomentum] = useState(false);
> > > <ScrollView
> > >       onScroll={( event ) => {
> > >            setCanMomentum(true)
> > >      }}
> > >      onMomentumScrollEnd={() => {
> > >            if (canmomentum) console.log('onMomentumScrollEnd')
> > >            setCanMomentum(false)
> > >      }}
> > > >
> > >      <Content/>
> > > </ScrollView>
> > > ```
> > 
> > 
> > Thanks for this suggestion, I did something similar but with a ref instead to avoid unnecessary state updates and renders.
> 
> I didn't find how to reference the scrolling state from ref. would you agree to share your solution?

Second this, please share thy precious knowledge of the hooks. Saveth me from mine own perils! @ewnavilae

**Comment by josecarlosrx** [comment]
`const canMomentum = React.useRef(false)

const onMomentumScrollBegin = () => {
    canMomentum.current = true;
};

const onMomentumScrollEnd = () => {

    if (canMomentum.current) {
        // console.log('onMomentumScrollEnd');
    }

    canMomentum.current = false;

};

<ScrollView onMomentumScrollBegin={onMomentumBegin} onMomentumScrollEnd={onMomentumScrollEnd}></ScrollView>`

**Comment by MarceloNascc** [comment]
I'm having the same issue in version 0.68.2

**Comment by RobbertMerapar** [comment]
Still an issue in version 0.69.6

**Comment by MichHighlander** [comment]
Still an issue in version 0.70

**Comment by stingerdallas** [comment]
Any update on this? The workaround doesn't work if using the snapToInterval because its the last event that is correct.  I can put in a debounce but it is way too long of a debounce since the onMomentumScrollEnd gets called almost a second apart.

When using snapToInterval, and user drags very little and lets go, it snaps back to the original snap location, but onMomentumScrollEnd gets called twice. once for the temp scrolled location and once for the new location

**Comment by kirx76** [comment]
Still an issue in version 0.71.7

**Comment by wyeo** [comment]
+1

**Comment by Jlexyc** [comment]
+1

**Comment by Bowlerr** [comment]
+1

**Comment by gaearon** [comment]
Looks like this will be fixed by https://github.com/facebook/react-native/pull/32433.

**Comment by gaearon** [comment]
Note there's a similar problem for `onScrollBeginDrag` (which gets called twice). I don't know if that PR fixes it.

**Comment by hichemfantar** [comment]
I'm still having this issue in react-native 0.72.6.

**Comment by hichemfantar** [comment]
> Solution using useRef:
> 
> ```
> const canMomentum = React.useRef(false);
> 
> const onMomentumScrollBegin = () => {
>   canMomentum.current = true;
> };
> 
> const onMomentumScrollEnd = () => {
>   if (canMomentum.current) {
>     // console.log('onMomentumScrollEnd');
>   }
> 
>   canMomentum.current = false;
> };
> 
> return (
>   <ScrollView
>     onMomentumScrollBegin={onMomentumScrollBegin}
>     onMomentumScrollEnd={onMomentumScrollEnd}
>   ></ScrollView>
> );
> ```

This works but the event is called too early, it doesn't wait for the animation to end.

**Comment by xaiamov** [comment]
It must be fixed in RN 0.74.1
https://github.com/facebook/react-native/commit/06668fcbacd750771f1d53cce829dc55e86f3f3c

**Comment by Iuriy-Budnikov** [comment]
Not fixed :(

**Comment by mars-lan** [comment]
Using RN 0.74.3 and can confirm that this issue has been fixed.

**Comment by hichemfantar** [comment]
looks like this is fixed, should probably be closed.

**Comment by hichemfantar** [comment]
actually it seems the behavior isn't consistent between android and ios.
on android it works when you flick and release, it also works when you scroll and hold thereby controlling the scrolling.

on IOS it only works when you flick and let the scroll momentum end by itself.

**Comment by hichemfantar** [comment]
I created a separate issue here: https://github.com/facebook/react-native/issues/48423
i believe this can be closed now since it's a different issue

**Comment by cortinico** [comment]
> I created a separate issue here: #48423 i believe this can be closed now since it's a different issue

Closing as per this
